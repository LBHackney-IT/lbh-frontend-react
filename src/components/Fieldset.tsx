import classNames from "classnames";
import PropTypes from "prop-types";
import React from "react";

import "lbh-frontend/lbh/components/lbh-fieldset/_fieldset.scss";

/**
 * The property types for {@link FieldsetLegend}
 */
export interface FieldsetLegendProps {
  id?: string;
  className?: string;
  isPageHeading?: boolean;
  children?: React.ReactNode;
}

/**
 * The type of a valid fieldset legend component. It allows any React functional
 * components that accept {@link FieldsetLegendProps} as proptypes.
 */
export type FieldsetLegendType = React.FunctionComponent<FieldsetLegendProps>;

/**
 * The legend for a fieldset
 */
export const FieldsetLegend: FieldsetLegendType = (
  props: FieldsetLegendProps
): JSX.Element => {
  const { id, className, isPageHeading, children } = props;
  return (
    <legend id={id} className={classNames("govuk-fieldset__legend", className)}>
      {isPageHeading ? (
        <h1 className="govuk-fieldset__heading">{children}</h1>
      ) : (
        children
      )}
    </legend>
  );
};

FieldsetLegend.propTypes = {
  id: PropTypes.string,
  className: PropTypes.string,
  isPageHeading: PropTypes.bool,
  children: PropTypes.node
};
