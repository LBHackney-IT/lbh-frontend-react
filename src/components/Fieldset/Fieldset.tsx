import classNames from "classnames";
import PropTypes from "prop-types";
import React from "react";
import { nullValuesAsUndefined } from "null-as-undefined";

import { Attributes, DataAttributes } from "../../helpers/Attributes";

import "lbh-frontend/lbh/components/lbh-fieldset/_fieldset.scss";

/**
 * The proptypes for {@link Fieldset}.
 *
 * This also supports all `aria-*` and `data-*` attributes.
 *
 * @noInheritDoc
 */
export interface FieldsetProps extends FieldsetPropsWithoutChildren {
  children: React.ReactNode;
}

export interface FieldsetPropsWithoutChildren
  extends React.AriaAttributes,
    DataAttributes {
  id?: string | null;
  className?: string | null;
  role?: string | null;
  /**
   * The legend for the fieldset. You should pass in a {@link FieldsetLegend}
   * component.
   */
  legend?: React.ReactNode | null;
  describedBy?: string | null;
}
/**
 * A component used to group elements within a form.
 */
export const Fieldset: React.FunctionComponent<FieldsetProps> = (
  props: FieldsetProps
): React.ReactElement => {
  const {
    id,
    className,
    legend,
    role,
    describedBy,
    children
  } = nullValuesAsUndefined(props);

  const extraAttributes = Attributes.ariaAndData(props);

  return (
    <fieldset
      id={id}
      className={classNames("govuk-fieldset lbh-fieldset", className)}
      role={role}
      {...extraAttributes}
      aria-describedby={describedBy}
    >
      {legend}
      {children}
    </fieldset>
  );
};

export const fieldsetWithoutChildrenPropTypes: PropTypes.ValidationMap<FieldsetPropsWithoutChildren> = {
  id: PropTypes.string,
  className: PropTypes.string,
  role: PropTypes.string,
  legend: PropTypes.element,
  describedBy: PropTypes.string
};

Fieldset.propTypes = {
  ...fieldsetWithoutChildrenPropTypes,
  children: PropTypes.node.isRequired
};
