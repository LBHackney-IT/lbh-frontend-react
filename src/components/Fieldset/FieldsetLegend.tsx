import classNames from "classnames";
import PropTypes from "prop-types";
import React from "react";

import { Attributes, DataAttributes } from "../../helpers/Attributes";

import "lbh-frontend/lbh/components/lbh-fieldset/_fieldset.scss";

/**
 * The proptypes for {@link FieldsetLegend}.
 *
 * This also supports all `aria-*` and `data-*` attributes.
 *
 * @noInheritDoc
 */
export interface FieldsetLegendProps
  extends React.AriaAttributes,
    DataAttributes {
  id?: string;
  className?: string;
  /**
   * Whether the legend also acts as the heading for the page. Places the legend
   * children inside a `h1` if set to `true`.
   */
  isPageHeading?: boolean;
  children: React.ReactNode;
}

/**
 * A component to set a legend for a {@link Fieldset} by passing
 * in via the {@link FieldsetProps.legend} prop.
 *
 * ```ts
 * // Creating a fieldset with a legend
 * <Fieldset
 *   id="fieldsetId"
 *   className="fieldsetClass"
 *   describedBy="fieldsetDescribedBy"
 *   role="fieldsetRole"
 *   legend={
 *     <FieldsetLegend id="legendId" className="legendClass" isPageHeading>
 *       children
 *     </FieldsetLegend>
 *   }
 * >
 *   children
 * </Fieldset>
 * ```
 */
export const FieldsetLegend: React.FunctionComponent<FieldsetLegendProps> = (
  props: FieldsetLegendProps
): JSX.Element => {
  const { id, className, isPageHeading, children } = props;

  const extraAttributes = Attributes.ariaAndData(props);

  return (
    <legend
      id={id}
      className={classNames(
        "govuk-fieldset__legend",
        "lbh-fieldset__legend",
        className
      )}
      {...extraAttributes}
    >
      {isPageHeading ? (
        <h1 className="govuk-fieldset__heading lbh-fieldset__heading">
          {children}
        </h1>
      ) : (
        children
      )}
    </legend>
  );
};

FieldsetLegend.propTypes = {
  id: PropTypes.string,
  className: PropTypes.string,
  isPageHeading: PropTypes.bool,
  children: PropTypes.node.isRequired
};
