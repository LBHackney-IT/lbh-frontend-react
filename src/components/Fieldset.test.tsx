import React from "react";
import { create } from "react-test-renderer";

import { FieldsetLegend } from "./Fieldset";

describe("FieldsetLegend", () => {
  it("renders correctly with all props", () => {
    const component = create(
      <FieldsetLegend id="Testid" className="class1 class2" isPageHeading>
        Test
      </FieldsetLegend>
    );

    expect(component).toMatchSnapshot();
  });

  it("renders correctly without optional props", () => {
    const component = create(<FieldsetLegend />);

    expect(component).toMatchSnapshot();
  });
});
