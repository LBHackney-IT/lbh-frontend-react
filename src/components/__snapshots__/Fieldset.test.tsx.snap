// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldsetLegend renders correctly with all props 1`] = `
<legend
  className="govuk-fieldset__legend class1 class2"
  id="Testid"
>
  <h1
    className="govuk-fieldset__heading"
  >
    Test
  </h1>
</legend>
`;

exports[`FieldsetLegend renders correctly without optional props 1`] = `
<legend
  className="govuk-fieldset__legend"
/>
`;
